File: src/services/api/bookmarks.ts
/**
 * Bookmark API 模块
 */
import { invoke } from "@tauri-apps/api/core";
import type { Bookmark } from "@/types";

export interface CreateBookmarkRequest {
  sourceId: string;
  position: string; // CFI 或等效位置标识
  label?: string;
  note?: string;
}

export interface UpdateBookmarkRequest {
  label?: string;
  note?: string;
  position?: string;
}

/**
 * 获取所有书签
 */
export async function getAll(): Promise<Bookmark[]> {
  return await invoke<Bookmark[]>("get_all_bookmarks");
}

/**
 * 获取指定文献源的书签
 */
export async function getBySource(sourceId: string): Promise<Bookmark[]> {
  return await invoke<Bookmark[]>("get_bookmarks_by_source", { sourceId });
}

/**
 * 获取单个书签
 */
export async function get(id: string): Promise<Bookmark | null> {
  return await invoke<Bookmark | null>("get_bookmark", { id });
}

/**
 * 创建书签
 */
export async function create(data: CreateBookmarkRequest): Promise<Bookmark> {
  return await invoke<Bookmark>("create_bookmark", { req: data });
}

/**
 * 更新书签
 */
export async function update(id: string, data: UpdateBookmarkRequest): Promise<Bookmark | null> {
  return await invoke<Bookmark | null>("update_bookmark", { id, req: data });
}

/**
 * 删除书签
 */
export async function deleteBookmark(id: string): Promise<void> {
  await invoke("delete_bookmark", { id });
}

// 导出 delete 别名
export { deleteBookmark as delete };

File: src-tauri/src/commands/bookmarks.rs
//! Bookmark 相关命令

use crate::models::{Bookmark, CreateBookmarkRequest, UpdateBookmarkRequest};
use crate::state::AppState;
use tauri::State;

/// 获取文献源的所有书签
#[tauri::command]
pub async fn get_bookmarks_by_source(state: State<'_, AppState>, source_id: String) -> Result<Vec<Bookmark>, String> {
    let services = state.get_services().ok_or("Vault not initialized")?;
    services
        .bookmark
        .get_by_source(&source_id)
        .await
        .map_err(|e| e.to_string())
}

/// 获取所有书签
#[tauri::command]
pub async fn get_all_bookmarks(state: State<'_, AppState>) -> Result<Vec<Bookmark>, String> {
    let services = state.get_services().ok_or("Vault not initialized")?;
    services.bookmark.get_all().await.map_err(|e| e.to_string())
}

/// 创建书签
#[tauri::command]
pub async fn create_bookmark(state: State<'_, AppState>, req: CreateBookmarkRequest) -> Result<Bookmark, String> {
    let services = state.get_services().ok_or("Vault not initialized")?;
    services.bookmark.create(req).await.map_err(|e| e.to_string())
}

/// 更新书签
#[tauri::command]
pub async fn update_bookmark(
    state: State<'_, AppState>,
    id: String,
    req: UpdateBookmarkRequest,
) -> Result<Option<Bookmark>, String> {
    let services = state.get_services().ok_or("Vault not initialized")?;
    services
        .bookmark
        .update(&id, req)
        .await
        .map_err(|e| e.to_string())
}

/// 删除书签
#[tauri::command]
pub async fn delete_bookmark(state: State<'_, AppState>, id: String) -> Result<(), String> {
    let services = state.get_services().ok_or("Vault not initialized")?;
    services.bookmark.delete(&id).await.map_err(|e| e.to_string())
}

/// 获取单个书签
#[tauri::command]
pub async fn get_bookmark(state: State<'_, AppState>, id: String) -> Result<Option<Bookmark>, String> {
    let services = state.get_services().ok_or("Vault not initialized")?;
    services.bookmark.get_by_id(&id).await.map_err(|e| e.to_string())
}




